File: bank-service/src/main/java/rs/edu/raf/service/impl/ExchangeRateServiceImpl.java
package rs.edu.raf.service.impl;

import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import rs.edu.raf.dto.ExchangeRateDto;
import rs.edu.raf.exceptions.ExchangeRateNotFoundException;
import rs.edu.raf.model.ExchangeRate;
import rs.edu.raf.mapper.ExchangeRateMapper;
import rs.edu.raf.repository.ExchangeRateRepository;
import rs.edu.raf.service.ExchangeRateService;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.List;
import java.util.Map;

@Service
    public class ExchangeRateServiceImpl implements ExchangeRateService {

    @Value("${exchange.api.url}")
    private String EXCHANGE_API_URL;
    private final ExchangeRateRepository exchangeRateRepository;
    private final ExchangeRateMapper exchangeRateMapper;

    public ExchangeRateServiceImpl(ExchangeRateRepository exchangeRateRepository, ExchangeRateMapper exchangeRateMapper) {
        this.exchangeRateRepository = exchangeRateRepository;
        this.exchangeRateMapper = exchangeRateMapper;
    }

    @PostConstruct
    public void saveExchangeRates(){
        RestTemplate restTemplate = new RestTemplate();
        Map<String, Object> exchangeRatesResponse = restTemplate.getForObject(EXCHANGE_API_URL, Map.class);

        if(exchangeRatesResponse != null && exchangeRatesResponse.containsKey("conversion_rates")){
            Map<String, Object> conversionRates = (Map<String, Object>) exchangeRatesResponse.get("conversion_rates");
            saveRates(conversionRates);
        }
    }

    private void saveRates(Map<String, Object> conversionRates){
    
        exchangeRateRepository.deleteAll();
        conversionRates.forEach((currencyCode, rate) -> {
            ExchangeRate exchangeRate = new ExchangeRate();
            exchangeRate.setCurrencyCode(currencyCode);
            exchangeRate.setRate(new BigDecimal(rate.toString()));
            exchangeRateRepository.save(exchangeRate);
        });
    }

    @Override
    public List<ExchangeRateDto> getAllExchangeRates(){
        return exchangeRateRepository.findAll()
                .stream()
                .map(exchangeRateMapper::exchangeRateToExchangeRateResponseDto)
                .toList();
    }

    @Override
    public BigDecimal convertToCurrency(String oldCurrencyCode, String newCurrencyCode, BigDecimal amount){

        ExchangeRate oldExchangeRate = exchangeRateRepository.findByCurrencyCode(oldCurrencyCode)
                .orElseThrow(()->new ExchangeRateNotFoundException(oldCurrencyCode));

        ExchangeRate newExchangeRate = exchangeRateRepository.findByCurrencyCode(newCurrencyCode)
                .orElseThrow(()->new ExchangeRateNotFoundException(newCurrencyCode));

        BigDecimal oldAmount = oldExchangeRate.getRate();
        BigDecimal newAmount = newExchangeRate.getRate();

        BigDecimal provision = BigDecimal.valueOf(0.995);

        return amount.divide(oldAmount, new MathContext(2, RoundingMode.HALF_DOWN)).multiply(newAmount).multiply(provision);
    }

    @Override
    public BigDecimal exchangeRate(String oldCurrencyCode, String newCurrencyCode){

        ExchangeRate oldExchangeRate = exchangeRateRepository.findByCurrencyCode(oldCurrencyCode)
                .orElseThrow(()->new ExchangeRateNotFoundException(oldCurrencyCode));

        ExchangeRate newExchangeRate = exchangeRateRepository.findByCurrencyCode(newCurrencyCode)
                .orElseThrow(()->new ExchangeRateNotFoundException(newCurrencyCode));

        BigDecimal oldAmount = oldExchangeRate.getRate();
        BigDecimal newAmount = newExchangeRate.getRate();

        return new BigDecimal("1.0").divide(oldAmount,new MathContext(2, RoundingMode.HALF_DOWN)).multiply(newAmount);
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/service/impl/ExchangeRateServiceImpl.kt
package org.example.bankservicekotlin.service.impl

import jakarta.annotation.PostConstruct
import org.example.bankservicekotlin.dto.ExchangeRateDTO
import org.example.bankservicekotlin.exception.ExchangeRateNotFoundException
import org.example.bankservicekotlin.mapper.ExchangeRateMapper
import org.example.bankservicekotlin.model.ExchangeRate
import org.example.bankservicekotlin.repository.ExchangeRateRepository
import org.example.bankservicekotlin.service.ExchangeRateService
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.getForObject
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

@Service
class ExchangeRateServiceImpl(
    @Value("\${exchange.api.url}")
    val EXCHANGE_API_URL: String,
    private val exchangeRateRepository: ExchangeRateRepository,
    private val exchangeRateMapper: ExchangeRateMapper
): ExchangeRateService {
    @PostConstruct
    fun saveExchangeRates() {
        val restTemplate = RestTemplate()
        val exchangeRatesResponse: Map<String, Any>? = restTemplate.getForObject(EXCHANGE_API_URL, Map::class)

        if (exchangeRatesResponse != null && exchangeRatesResponse.containsKey("conversion_rates")) {
            val conversionRates = exchangeRatesResponse["conversion_rates"] as Map<String, Any>
            saveRates(conversionRates)
        }
    }

    private fun saveRates(conversionRates: Map<String, Any>) {
        exchangeRateRepository.deleteAll()
        conversionRates.forEach { (currencyCode: String, rate: Any) ->
            val exchangeRate = ExchangeRate(null, currencyCode, BigDecimal(rate.toString()))
            exchangeRateRepository.save(exchangeRate)
        }
    }

    override fun getAllExchangeRates(): List<ExchangeRateDTO> {
        return exchangeRateRepository.findAll()
            .stream()
            .map(exchangeRateMapper::exchangeRateToExchangeRateResponseDto)
            .toList()
    }

    override fun convertToCurrency(oldCurrencyCode: String, newCurrencyCode: String, amount: BigDecimal): BigDecimal {
        val oldExchangeRate: ExchangeRate = exchangeRateRepository.findByCurrencyCode(oldCurrencyCode)
            .orElseThrow { ExchangeRateNotFoundException(oldCurrencyCode) }

        val newExchangeRate: ExchangeRate = exchangeRateRepository.findByCurrencyCode(newCurrencyCode)
            .orElseThrow { ExchangeRateNotFoundException(newCurrencyCode) }

        val oldAmount: BigDecimal = oldExchangeRate.rate
        val newAmount: BigDecimal = newExchangeRate.rate

        val provision = BigDecimal.valueOf(0.995)

        return amount.divide(oldAmount, MathContext(2, RoundingMode.HALF_DOWN)).multiply(newAmount).multiply(provision)
    }

    override fun exchangeRate(oldCurrencyCode: String, newCurrencyCode: String): BigDecimal {
        val oldExchangeRate: ExchangeRate = exchangeRateRepository.findByCurrencyCode(oldCurrencyCode)
            .orElseThrow { ExchangeRateNotFoundException(oldCurrencyCode) }

        val newExchangeRate: ExchangeRate = exchangeRateRepository.findByCurrencyCode(newCurrencyCode)
            .orElseThrow { ExchangeRateNotFoundException(newCurrencyCode) }

        val oldAmount: BigDecimal = oldExchangeRate.rate
        val newAmount: BigDecimal = newExchangeRate.rate

        return BigDecimal("1.0").divide(oldAmount, MathContext(2, RoundingMode.HALF_DOWN)).multiply(newAmount)
    }
}